sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java

```diff
@@ -1045,4 +1045,47 @@ private boolean validateNoETag(BlobRequestConditions modifiedRequestConditions)
         }
         return modifiedRequestConditions.getIfMatch() == null && modifiedRequestConditions.getIfNoneMatch() == null;
     }
+
+    /**
+     * Restores a previously deleted container.  The restored container
+     * will be renamed to the name of this {@link BlobContainerAsyncClient}.
+     * If the container associated with this {@link BlobContainerAsyncClient}
+     * already exists, this call will result in a 409 (conflict).
+     * This API is only functional if Container Soft Delete is enabled
+     * for the storage account associated with the container.
+     *
+     * @param deletedContainerName The name of the previously deleted container.
+     * @param deletedContainerVersion The version of the previously deleted container.
+     * @return A reactive response signaling completion.
+     */
+    public Mono<Void> restore(String deletedContainerName, String deletedContainerVersion) {
```

Comment:
Was there a discussion around restore vs undelete? Restore sounds cleaner, but we already have undelete blobs, so not sure if we want to be consistent.

Reply:
This also differs from blobs in that with blobs the client must point to the deleted blob. Here the client points to the new destination. It may be worth discussing whether we depart from the rest api and have the client point to the deleted container and then do some work internally to switch things around.

Reply:
@rickle-msft  We have a discussion started above - https://github.com/Azure/azure-sdk-for-java/pull/10900#discussion_r421834330 - please see ideas there and share your pick.

Reply:
We landed on "undelete" and moving up to service client.